source/extensions/common/wasm/wasm_vm_base.h

```diff
@@ -0,0 +1,61 @@
+#pragma once
+
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "extensions/common/wasm/wasm_vm.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+
+/**
+ * Wasm host stats.
+ */
+#define ALL_VM_STATS(COUNTER, GAUGE)                                                               \
+  COUNTER(created)                                                                                 \
+  COUNTER(cloned)                                                                                  \
+  GAUGE(active, NeverImport)
+
+struct VmStats {
+  ALL_VM_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT)
+};
+
+// Wasm VM base instance. Provides common behavior (e.g. Stats).
+class WasmVmBase : public WasmVm {
+public:
+  WasmVmBase(Stats::ScopeSharedPtr scope, std::atomic<int>* active_vms_ptr,
+             absl::string_view runtime)
+      : scope_(scope), active_vms_ptr_(active_vms_ptr),
+        runtime_prefix_(absl::StrCat("wasm_vm.", runtime, ".")),
+        stats_(VmStats{ALL_VM_STATS(POOL_COUNTER_PREFIX(*scope_, runtime_prefix_),
+                                    POOL_GAUGE_PREFIX(*scope_, runtime_prefix_))}),
+        runtime_(std::string(runtime)) {
+    stats_.created_.inc();
+    (*active_vms_ptr_)++;
+    stats_.active_.set(*active_vms_ptr_);
```

Comment:
If WasmVmBase can be constructed from multiple threads, then I think another thread could come and bump active_vms_ptr_ between the above and below lines.

So I don't think these 2 atomics can be updated atomically like this. You should use a mutex for these.

If WasmVmBase is only constructed from the main thread, then you don't need active_vms_ptr_ to be an atomic at all.

Reply:
It is only on the main thread, but I would agree that that it too subtle.   Changed the code to store the result of the atomic operations and use that on subsequent lines.  PTAL